Enable collectors to take advantage of pre-aggregated data.

jpountz opened the pull request:
This introduces `LeafCollector#collectRange`, which is typically useful to take advantage of the pre-aggregated data exposed in `DocValuesSkipper`. At the moment, `DocValuesSkipper` only exposes per-block min and max values, but we could easily extend it to record sums and value counts as well.

This `collectRange` method would be called if there are no deletions in the segment by:
 - queries that rewrite to a `MatchAllDocsQuery` (with min=0 and max=maxDoc),
 - `PointRangeQuery` on segments that fully match the range (typical for time-based data),
 - doc-value range queries and conjunctions of doc-value range queries on fields that enable sparse indexing and correlate with the index sort.

jpountz committed:
Enable collectors to take advantage of pre-aggregated data.

This introduces `LeafCollector#collectRange`, which is typically useful to take
advantage of the pre-aggregated data exposed in `DocValuesSkipper`. At the
moment, `DocValuesSkipper` only exposes per-block min and max values, but we
could easily extend it to record sums and value counts as well.

This `collectRange` method would be called if there are no deletions in the
segment by:
 - queries that rewrite to a `MatchAllDocsQuery` (with min=0 and max=maxDoc),
 - `PointRangeQuery` on segments that fully match the range (typical for
   time-based data),
 - doc-value range queries and conjunctions of doc-value range queries on
   fields that enable sparse indexing and correlate with the index sort.

jpountz commented:
@epotyom You may be interested in this, this allows computing aggregates in sub-linear time respective to the number of matching docs.

gsmiller commented:
It makes sense to me to expose the idea of doc range collection as a first-class API on leaf collectors for the reasons you outlined above. This would also benefit #14273 as well right? If there are scorers that can leverage the range collection call, it would immediately benefit I believe.

jpountz commented:
> This would also benefit https://github.com/apache/lucene/pull/14273

I don't think so, or rather taking advantage of range collection shouldn't help more than what #14273 does with `RangeDocIdStream`?

For clarity, this `collectRange` method is more useful for aggregating values (facet "recorders" if I use the terminology of the new faceting framework?). Implementations would need to consult the `DocValuesSkipper` to check if it has pre-aggregated data over ranges of doc IDs that are contained in the range of doc IDs passed to `collectRange`. These sub ranges could be aggregated in constant time, without having to iterate over docs.

jpountz commented on lucene/core/src/java/org/apache/lucene/search/RangeDocIdStream.java:
```diff
@@ -0,0 +1,44 @@
+  private final int min, max;
+
+  RangeDocIdStream(int min, int max) {
+    if (max < min) {
```
This is a good question, I had overlooked it. While I don't think that empty ranges would cause issues for any impl, I think we should reject them. I'll update the PR.

gsmiller commented:
> I don't think so, or rather taking advantage of range collection shouldn't help more than what https://github.com/apache/lucene/pull/14273 does with RangeDocIdStream?

My thinking here was that `HistogramCollector` should benefit from any scorers that can provide it with a `DocIdStream` for collection, and that this change lays the groundwork for more scorers to pass streams to collectors instead of individual docs (specifically thinking about some of the query use-cases you mention in the description). I probably should have been more clear :) (and maybe I'm still getting confused and this isn't true...)

jpountz committed:
Merge branch 'main' into collect_pre_aggregated_data

jpountz commented:
Ah, that's right. We have a good number of queries that are already covered, in my opinion the next natural step is to look into making ranges collect ranges when any clause would collect a range.

jpountz commented:
Any opinion on `collect(int min, int max)` vs. `collectRange(int min, int max)`? I leaned towards `collectRange` since we already have `collect(int doc)` and it wouldn't be obvious from the parameter types whether `collect(int, int)` is collecting a range or two random docs. No strong feeling either way though. `collect(DocIdStream)` is called "collect" rather than "collectDocIdStream" so I guess that `collect(int min, int max)` would be more consistent from this perspective.

gsmiller commented:
I prefer `collectRange` as well to make usage a little less error-prone. I don't have a strong opinion though.

jpountz committed:
Merge branch 'main' into collect_pre_aggregated_data

jpountz committed:
Review feedback

jpountz committed:
Fix HistogramCollector to not create zero-count buckets.

If a bucket in the middle of the range doesn't match docs, it would be returned
with a count of zero. Better not return it at all.

gf2121 commented on lucene/test-framework/src/java/org/apache/lucene/tests/search/AssertingLeafCollector.java:
```diff
@@ -50,6 +50,14 @@ public void collect(DocIdStream stream) throws IOException {
+  @Override
+  public void collectRange(int min, int max) throws IOException {
+    assert min > lastCollected;
+    assert max > min;
```
Maybe assert `min >= this.min` and `max <= this.max` as well :)

gf2121 reviewed (approved):
LGTM

jpountz committed:
Merge branch 'main' into collect_pre_aggregated_data

jpountz committed:
Review feedback.

gf2121 reviewed (approved):
We also need a CHANGES entry.

gf2121 commented:
The change of https://github.com/apache/lucene/pull/14421 is also included, which seems not expected?

jpountz commented:
It is unexpected indeed! I'll fix this and add a CHANGES entry.

jpountz committed:
Merge branch 'main' into collect_pre_aggregated_data

jpountz committed:
CHANGES
